pkg/volume/flexvolume/flexvolume.go

```diff
@@ -91,236 +94,291 @@ func (plugin *flexVolumePlugin) GetAccessModes() []api.PersistentVolumeAccessMod
 	}
 }
 
-func (plugin *flexVolumePlugin) getVolumeSource(spec *volume.Spec) *api.FlexVolumeSource {
-	var source *api.FlexVolumeSource
-	if spec.Volume != nil && spec.Volume.FlexVolume != nil {
-		source = spec.Volume.FlexVolume
-	} else if spec.PersistentVolume != nil && spec.PersistentVolume.Spec.FlexVolume != nil {
-		source = spec.PersistentVolume.Spec.FlexVolume
+// Attacher interface
+
+func (plugin *flexVolumePlugin) Attach(spec *volume.Spec, hostName string) error {
+	call := NewDriverCall(plugin.getExecutable(), attachCmd, exec.New())
+	call.AppendSpec(spec, plugin.host)
+	call.Append(hostName)
+
+	_, err := call.Run()
+	if err != nil && isCmdNotSupportedErr(err) {
+		return nil
 	}
-	return source
+	return err
 }
 
-// NewMounter is the mounter routine to build the volume.
-func (plugin *flexVolumePlugin) NewMounter(spec *volume.Spec, pod *api.Pod, _ volume.VolumeOptions) (volume.Mounter, error) {
-	fv := plugin.getVolumeSource(spec)
+func (plugin *flexVolumePlugin) WaitForAttach(spec *volume.Spec, timeout time.Duration) (string, error) {
+	call := NewDriverCallWithTimeout(plugin.getExecutable(), waitForAttachCmd, exec.New(), timeout)
+	call.AppendSpec(spec, plugin.host)
+
+	status, err := call.Run()
+	if err != nil {
+		return "", err
+	}
+	return status.Device, nil
+}
+
+func (plugin *flexVolumePlugin) GetDeviceMountPath(spec *volume.Spec) string {
+	call := NewDriverCall(plugin.getExecutable(), getDeviceMountPathCmd, exec.New())
+	call.AppendSpec(spec, plugin.host)
+
+	status, err := call.Run()
+	if isCmdNotSupportedErr(err) {
+		name := plugin.driverName
+		return plugin.host.GetPodVolumeDir(spec.PodUID, utilstrings.EscapeQualifiedNameForDisk(name), spec.Name())
+	} else if err != nil {
+		glog.Errorf("FlexVolume: driver bug in %s: returned an error on getdevicemountpath: %s", plugin.getExecutable(), err)
+		return ""
+	}
+	return status.Path
+}
+
+func (plugin *flexVolumePlugin) MountDevice(spec *volume.Spec, devicePath string, deviceMountPath string, mounter mount.Interface) error {
+	call := NewDriverCall(plugin.getExecutable(), mountDeviceCmd, exec.New())
+	call.AppendSpec(spec, plugin.host)
```

Comment:
We need to create and pass options like fsType, read write.
Please also add a new option "kubernetes.io/fsgroup" for folks willing to implement their own mount and set fsgroup.